File: frontend/src/api/training.ts
import http from './http'


export type SelectOption = {
  label: string
  value: string
  color_type?: string | null
  css_class?: string | null
}


export type TrainingStudentOption = {
  student_no: string
  student_name: string
  advisor_name: string
  label: string
}


export type TrainingPlanRecord = {
  id: number
  student_no: string
  student_name: string
  advisor_name: string
  version_no: string
  report_cycle: string
  plan_status: string
  scientific_goal: string
  assessment_rule: string
}


export type TrainingPlanUpsert = Omit<TrainingPlanRecord, 'id'>


export type ScientificReportRecord = {
  id: number
  student_no: string
  student_name: string
  period_label: string
  report_status: string
  reviewer_name?: string | null
  review_score?: number | null
  summary: string
}


export type ScientificReportUpsert = Omit<ScientificReportRecord, 'id'>


export type OutboundStudyRecord = {
  id: number
  student_no: string
  student_name: string
  advisor_name: string
  study_type: string
  destination: string
  start_date: string
  end_date: string
  approval_status: string
  expected_outcome?: string | null
}


export type OutboundStudyUpsert = Omit<OutboundStudyRecord, 'id'>


export type TrainingOptions = {
  plan_status_options: SelectOption[]
  report_cycle_options: SelectOption[]
  report_status_options: SelectOption[]
  study_type_options: SelectOption[]
  approval_status_options: SelectOption[]
  advisor_options: SelectOption[]
  reviewer_options: SelectOption[]
  student_options: TrainingStudentOption[]
}


export type TrainingStats = {
  training_plan_total: number
  pending_confirmation_total: number
  report_pending_total: number
  outbound_active_total: number
}


export function getTrainingStats() {
  return http.get<TrainingStats>('/training/stats')
}


export function getTrainingOptions() {
  return http.get<TrainingOptions>('/training/options')
}


export function listTrainingPlans(params?: { keyword?: string; plan_status?: string; advisor_name?: string; report_cycle?: string }) {
  return http.get<{ items: TrainingPlanRecord[]; total: number }>('/training/plans', { params })
}


export function createTrainingPlan(payload: TrainingPlanUpsert) {
  return http.post<TrainingPlanRecord>('/training/plans', payload)
}


export function updateTrainingPlan(id: number, payload: TrainingPlanUpsert) {
  return http.put<TrainingPlanRecord>(`/training/plans/${id}`, payload)
}


export function deleteTrainingPlan(id: number) {
  return http.delete(`/training/plans/${id}`)
}


export function batchDeleteTrainingPlans(ids: number[]) {
  return http.post<{ success_count: number }>('/training/plans/batch-delete', { ids })
}


export function listScientificReports(params?: { keyword?: string; status?: string; reviewer_name?: string }) {
  return http.get<{ items: ScientificReportRecord[]; total: number }>('/training/reports', { params })
}


export function createScientificReport(payload: ScientificReportUpsert) {
  return http.post<ScientificReportRecord>('/training/reports', payload)
}


export function updateScientificReport(id: number, payload: ScientificReportUpsert) {
  return http.put<ScientificReportRecord>(`/training/reports/${id}`, payload)
}


export function deleteScientificReport(id: number) {
  return http.delete(`/training/reports/${id}`)
}


export function batchDeleteScientificReports(ids: number[]) {
  return http.post<{ success_count: number }>('/training/reports/batch-delete', { ids })
}


export function listOutboundStudies(params?: { keyword?: string; status?: string; study_type?: string; advisor_name?: string }) {
  return http.get<{ items: OutboundStudyRecord[]; total: number }>('/training/outbound-studies', { params })
}


export function createOutboundStudy(payload: OutboundStudyUpsert) {
  return http.post<OutboundStudyRecord>('/training/outbound-studies', payload)
}


export function updateOutboundStudy(id: number, payload: OutboundStudyUpsert) {
  return http.put<OutboundStudyRecord>(`/training/outbound-studies/${id}`, payload)
}


export function deleteOutboundStudy(id: number) {
  return http.delete(`/training/outbound-studies/${id}`)
}


export function batchDeleteOutboundStudies(ids: number[]) {
  return http.post<{ success_count: number }>('/training/outbound-studies/batch-delete', { ids })
}

File: backend/app/schemas/training.py
from pydantic import BaseModel

from app.schemas.system import SelectOption


class TrainingTask(BaseModel):
    title: str
    owner: str
    due_text: str
    status: str


class TrainingWorkbench(BaseModel):
    open_tasks: list[TrainingTask]
    supervision_rules: list[dict[str, str]]
    outbound_study_status: list[dict[str, str | int]]


class DegreeWorkbench(BaseModel):
    thesis_pipeline: list[dict[str, str | int]]
    committee_tasks: list[TrainingTask]


class TrainingPlanRecord(BaseModel):
    id: int
    student_no: str
    student_name: str
    advisor_name: str
    version_no: str
    report_cycle: str
    plan_status: str
    scientific_goal: str
    assessment_rule: str


class TrainingPlanUpsert(BaseModel):
    student_no: str
    student_name: str
    advisor_name: str
    version_no: str
    report_cycle: str
    plan_status: str
    scientific_goal: str
    assessment_rule: str


class TrainingPlanListResponse(BaseModel):
    items: list[TrainingPlanRecord]
    total: int


class ScientificReportRecord(BaseModel):
    id: int
    student_no: str
    student_name: str
    period_label: str
    report_status: str
    reviewer_name: str | None = None
    review_score: float | None = None
    summary: str


class ScientificReportUpsert(BaseModel):
    student_no: str
    student_name: str
    period_label: str
    report_status: str
    reviewer_name: str | None = None
    review_score: float | None = None
    summary: str


class ScientificReportListResponse(BaseModel):
    items: list[ScientificReportRecord]
    total: int


class OutboundStudyRecord(BaseModel):
    id: int
    student_no: str
    student_name: str
    advisor_name: str
    study_type: str
    destination: str
    start_date: str
    end_date: str
    approval_status: str
    expected_outcome: str | None = None


class OutboundStudyUpsert(BaseModel):
    student_no: str
    student_name: str
    advisor_name: str
    study_type: str
    destination: str
    start_date: str
    end_date: str
    approval_status: str
    expected_outcome: str | None = None


class OutboundStudyListResponse(BaseModel):
    items: list[OutboundStudyRecord]
    total: int


class TrainingStudentOption(BaseModel):
    student_no: str
    student_name: str
    advisor_name: str
    label: str


class TrainingOptionsResponse(BaseModel):
    plan_status_options: list[SelectOption]
    report_cycle_options: list[SelectOption]
    report_status_options: list[SelectOption]
    study_type_options: list[SelectOption]
    approval_status_options: list[SelectOption]
    advisor_options: list[SelectOption]
    reviewer_options: list[SelectOption]
    student_options: list[TrainingStudentOption]


class TrainingStats(BaseModel):
    training_plan_total: int
    pending_confirmation_total: int
    report_pending_total: int
    outbound_active_total: int


class ThesisRecord(BaseModel):
    id: int
    student_no: str
    student_name: str
    advisor_name: str
    title: str
    plagiarism_rate: float | None = None
    thesis_status: str
    blind_review_status: str
    defense_status: str
    degree_status: str


class ThesisUpsert(BaseModel):
    student_no: str
    student_name: str
    advisor_name: str
    title: str
    plagiarism_rate: float | None = None
    thesis_status: str
    blind_review_status: str
    defense_status: str
    degree_status: str


class ThesisListResponse(BaseModel):
    items: list[ThesisRecord]
    total: int


class ThesisReviewRecord(BaseModel):
    id: int
    thesis_id: int
    thesis_title: str
    expert_name: str
    review_score: float | None = None
    review_status: str
    review_comment: str | None = None


class ThesisReviewUpsert(BaseModel):
    thesis_id: int
    thesis_title: str
    expert_name: str
    review_score: float | None = None
    review_status: str
    review_comment: str | None = None


class ThesisReviewListResponse(BaseModel):
    items: list[ThesisReviewRecord]
    total: int


class DegreeOptionsResponse(BaseModel):
    student_options: list[SelectOption]
    advisor_options: list[SelectOption]
    thesis_options: list[SelectOption]
    thesis_status_options: list[SelectOption]
    blind_review_status_options: list[SelectOption]
    defense_status_options: list[SelectOption]
    degree_status_options: list[SelectOption]
    expert_options: list[SelectOption]
    review_status_options: list[SelectOption]


class DegreeStats(BaseModel):
    thesis_total: int
    plagiarism_pending_total: int
    blind_review_pending_total: int
    defense_pending_total: int
    degree_granted_total: int
